File: 301_Remove_Invalid_Parentheses/301_2.cpp
class Solution {
public:
    vector<string> removeInvalidParentheses(string s) {
        unordered_set<string> res;
        string tmp;
        int left = 0, right = 0;
        for (char c : s) {
            if (c == '(') {
                left++;
            } else if (c == ')') {
                if (left > 0) left--;
                else          right++;
            }
        }
        dfs(s, tmp, 0, 0, 0, left, right, res);
        return vector<string>(res.begin(), res.end());
    }

    void dfs(string s, string &tmp, int idx, int leftKeep, int rightKeep, int left, int right, unordered_set<string> &res) {
        if (idx == s.size()) {
            if (left == 0 && right == 0) {
                res.insert(tmp);
            }
            return;
        }
        if (s[idx] == '(' && left > 0) {
            dfs(s, tmp, idx + 1, leftKeep, rightKeep, left - 1, right, res);
        }
        if (s[idx] == ')' && right > 0) {
            dfs(s, tmp, idx + 1, leftKeep, rightKeep, left, right - 1, res);
        }
        tmp.push_back(s[idx]);
        if (s[idx] != '(' && s[idx] != ')') {
            dfs(s, tmp, idx + 1, leftKeep, rightKeep, left, right, res);
        } else if (s[idx] == '(') {
            dfs(s, tmp, idx + 1, leftKeep + 1, rightKeep, left, right, res);
        } else if (leftKeep > rightKeep) {
            dfs(s, tmp, idx + 1, leftKeep, rightKeep + 1, left, right, res);
        }
        tmp.pop_back();
    }
};

File: 301_Remove_Invalid_Parentheses/301_2.py
class Solution:
    def removeInvalidParentheses(self, s: str) -> List[str]:
        def isValid(string):
            count = 0
            for i in string:
                if i == '(':
                    count += 1
                elif i == ')':
                    count -= 1
                if count < 0:
                    return False
            return count == 0
        
        def dfs(string, index, left, right):
            if left == 0 and right == 0 and isValid(string):
                res.append(string)
                return
            for i in range(index, len(string)):
                if i > index and string[i] == string[i-1]:
                    continue
                tmp = string[:i] + string[i+1:]
                if left > 0 and string[i] == '(':
                    dfs(tmp, i, left - 1, right)
                if right > 0 and string[i] == ')':
                    dfs(tmp, i, left, right - 1)
        
        res = []
        left, right = 0, 0
        for string in s:
            if string == '(':
                left += 1
            elif string == ')':
                if left > 0:
                    left -= 1
                else:
                    right += 1
        dfs(s, 0, left, right)
        return res
